core/serialisation/data_serialiser.py

```diff
@@ -0,0 +1,333 @@
+from sqlalchemy.orm import Query
+
+from core.db.entities import (
+    Funding,
+    FundingComment,
+    FundingQuestion,
+    Organisation,
+    OutcomeData,
+    OutcomeDim,
+    OutputData,
+    OutputDim,
+    PlaceDetail,
+    PrivateInvestment,
+    Programme,
+    ProgrammeProgress,
+    Project,
+    ProjectProgress,
+    RiskRegister,
+    Submission,
+)
+from core.db.queries import (
+    funding_comment_query,
+    funding_query,
+    funding_question_query,
+    organisation_query,
+    outcome_data_query,
+    outcome_dim_query,
+    output_data_query,
+    output_dim_query,
+    place_detail_query,
+    private_investment_query,
+    programme_progress_query,
+    programme_query,
+    project_progress_query,
+    project_query,
+    risk_register_query,
+)
+
+
+def serialise_download_data(base_query: Query, sheets_required: list[str] | None = None) -> dict[str, list[dict]]:
+    """
+    Query and serialise data from multiple tables for download.
+
+    Extend base query to return relevant fields for each table, and serialise accordingly. Calls individual
+    query methods and Marshmallow schema serialisers for each table, based on method names in table_queries dict.
+    All data added to dictionary.
+
+    :param base_query: An SQLAlchemy Query of core tables with filters applied.
+    :param sheets_required: Optional. List of sheets to query/serialise/return.
+    :yield: A tuple containing table name and serialised data.
+    """
+
+    table_queries = {
+        "PlaceDetails": (place_detail_query, PlaceDetailSchema),
+        "ProjectDetails": (project_query, ProjectSchema),
+        "OrganisationRef": (organisation_query, OrganisationSchema),
+        "ProgrammeRef": (programme_query, ProgrammeSchema),
+        "ProgrammeProgress": (programme_progress_query, ProgrammeProgressSchema),
+        "ProjectProgress": (project_progress_query, ProjectProgressSchema),
+        "FundingQuestions": (funding_question_query, FundingQuestionSchema),
+        "Funding": (funding_query, FundingSchema),
+        "FundingComments": (funding_comment_query, FundingCommentSchema),
+        "PrivateInvestments": (private_investment_query, PrivateInvestmentSchema),
+        "OutputsRef": (output_dim_query, OutputDimSchema),
+        "OutputData": (output_data_query, OutputDataSchema),
+        "OutcomeRef": (outcome_dim_query, OutcomeDimSchema),
+        "OutcomeData": (outcome_data_query, OutcomeDataSchema),
+        "RiskRegister": (risk_register_query, RiskRegisterSchema),
+    }
+
+    sheets_required = sheets_required if sheets_required else list(table_queries.keys())
+
+    serialised_results = {}
+
+    for sheet in sheets_required:
+        query_extender, schema = table_queries[sheet]
+        extended_query = query_extender(base_query)
+        table_data = extended_query.all()
+        table_serialised = schema(many=True).dump(table_data)
+        serialised_results[sheet] = table_serialised
```

Comment:
Could this be a comprehension rather than mutating the dict?

Reply:
I think in cases like this, you loose a lot of readability for limited performance gain, if switching to a comprehension.  After all it's only ever going to be doing like 14 iterations. Since there are quite a few lines executed for each iteration, this stuff can get unnecessarily complex in a comprehension. I'd prefer to keep it nice and verbose in a for loop, then we can easily follow it's steps, which also makes it easier to debug, and extend/change any of the functionality (for instance I may change it to a generator, depending on performance testing results)